File: checkout_flutter_bridge/android/src/main/kotlin/com/checkout/flutter/CheckoutFlutterBridgePlugin.kt
package com.checkout.flutter

import android.util.Log
import androidx.activity.ComponentActivity
import com.checkout.flutter.views.CardViewFactory
import com.checkout.flutter.views.GooglePayViewFactory
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * CheckoutFlutterBridgePlugin - Main plugin class Registers platform views and handles method
 * channel communication
 */
class CheckoutFlutterBridgePlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private var flutterPluginBinding: FlutterPlugin.FlutterPluginBinding? = null
    private var activity: ComponentActivity? = null
    private var cardPlatformView: CardPlatformView? = null
    private var googlePayPlatformView: GooglePayPlatformView? = null

    companion object {
        private const val CHANNEL = "checkout_bridge"
        private const val TAG = "CheckoutPlugin"
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        Log.d(TAG, "Plugin attached to engine")

        // Store the binding for later use when activity is attached
        this.flutterPluginBinding = flutterPluginBinding

        val messenger = flutterPluginBinding.binaryMessenger
        channel = MethodChannel(messenger, CHANNEL)
        channel.setMethodCallHandler(this)

        // Platform views will be registered when activity is attached
        Log.d(TAG, "Waiting for activity to register platform views")
    }

    private fun registerPlatformViews() {
        val binding = flutterPluginBinding ?: return
        val currentActivity = activity ?: return

        Log.d(TAG, "Registering platform views with activity")

        // Register platform views
        binding.platformViewRegistry.registerViewFactory(
                "flow_card_view",
                CardViewFactory(binding.binaryMessenger, currentActivity) { view ->
                    cardPlatformView = view
                    Log.d(TAG, "Card view instance captured")
                }
        )

        binding.platformViewRegistry.registerViewFactory(
                "flow_googlepay_view",
                GooglePayViewFactory(binding.binaryMessenger, currentActivity) { view ->
                    googlePayPlatformView = view
                    Log.d(TAG, "Google Pay view instance captured")
                }
        )

        Log.d(TAG, "Platform views registered successfully")
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        Log.d(TAG, "Method call: ${call.method}")

        when (call.method) {
            // ==================== CARD METHODS ====================
            "initCardView" -> {
                // Card view initialization happens in PlatformView
                result.success(true)
            }
            "initStoredCardView" -> {
                // Stored card view initialization happens in PlatformView
                result.success(true)
            }
            "validateCard" -> {
                if (cardPlatformView == null) {
                    result.error("CARD_NOT_READY", "Card view not initialized", null)
                    return
                }
                val isValid = cardPlatformView!!.validateCard()
                result.success(isValid)
            }
            "tokenizeCard" -> {
                if (cardPlatformView == null) {
                    result.error("CARD_NOT_READY", "Card view not initialized", null)
                    return
                }
                cardPlatformView?.tokenizeCard(result)
            }
            "getSessionData" -> {
                if (cardPlatformView == null) {
                    result.error("CARD_NOT_READY", "Card view not initialized", null)
                    return
                }
                cardPlatformView?.getSessionData(result)
            }

            // ==================== GOOGLE PAY METHODS ====================
            "initGooglePay" -> {
                // Google Pay initialization happens in PlatformView
                result.success(true)
            }
            "getGooglePaySessionData" -> {
                // Session data is automatically sent via handleSubmit callback
                result.success(true)
            }
            "tokenizeGooglePay" -> {
                if (googlePayPlatformView == null) {
                    result.error("GOOGLEPAY_NOT_READY", "Google Pay view not initialized", null)
                    return
                }
                googlePayPlatformView?.tokenizeGooglePay(result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        Log.d(TAG, "Plugin detached from engine")
        channel.setMethodCallHandler(null)
        cleanup()
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        Log.d(TAG, "Plugin attached to activity")
        activity =
                binding.activity as? ComponentActivity
                        ?: throw IllegalStateException("Activity must be a ComponentActivity")

        // Now that we have the activity, register platform views
        registerPlatformViews()
    }

    override fun onDetachedFromActivityForConfigChanges() {
        Log.d(TAG, "Plugin detached from activity for config changes")
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        Log.d(TAG, "Plugin reattached to activity after config changes")
        activity = binding.activity as? ComponentActivity
    }

    override fun onDetachedFromActivity() {
        Log.d(TAG, "Plugin detached from activity")
        activity = null
        cleanup()
    }

    private fun cleanup() {
        try {
            Log.d(TAG, "Cleaning up resources")
            cardPlatformView?.dispose()
            googlePayPlatformView?.dispose()

            cardPlatformView = null
            googlePayPlatformView = null
            flutterPluginBinding = null
        } catch (e: Exception) {
            Log.e(TAG, "Error during cleanup", e)
        }
    }
}

File: checkout_flutter_bridge/ios/Classes/CheckoutFlutterBridgePlugin.swift
import Flutter
import UIKit

public class CheckoutFlutterBridgePlugin: NSObject, FlutterPlugin {
    private static let channelName = "checkout_bridge"
    private var channel: FlutterMethodChannel?
    
    // Weak references to platform views
    private weak var cardPlatformView: CardPlatformView?
    private weak var applePayPlatformView: ApplePayPlatformView?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: channelName,
            binaryMessenger: registrar.messenger()
        )
        let instance = CheckoutFlutterBridgePlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        
        // Register platform view factories
        registerPlatformViews(with: registrar)
    }
    
    private static func registerPlatformViews(with registrar: FlutterPluginRegistrar) {
        // Register Card View Factory
        let cardViewFactory = CardViewFactory(messenger: registrar.messenger())
        registrar.register(
            cardViewFactory,
            withId: "flow_card_view"
        )
        
        // Register Apple Pay View Factory
        let applePayViewFactory = ApplePayViewFactory(messenger: registrar.messenger())
        registrar.register(
            applePayViewFactory,
            withId: "flow_applepay_view"
        )
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        // ==================== CARD METHODS ====================
        case "initCardView":
            // Card view initialization happens in PlatformView
            result(true)
            
        case "initStoredCardView":
            // Stored card view initialization happens in PlatformView
            result(true)
            
        case "validateCard":
            guard let cardView = cardPlatformView else {
                result(FlutterError(
                    code: "CARD_NOT_READY",
                    message: "Card view not initialized",
                    details: nil
                ))
                return
            }
            
            Task {
                let isValid = await cardView.validateCard()
                await MainActor.run {
                    result(isValid)
                }
            }
            
        case "tokenizeCard":
            guard let cardView = cardPlatformView else {
                result(FlutterError(
                    code: "CARD_NOT_READY",
                    message: "Card view not initialized",
                    details: nil
                ))
                return
            }
            cardView.tokenizeCard(result: result)
            
        case "getSessionData":
            guard let cardView = cardPlatformView else {
                result(FlutterError(
                    code: "CARD_NOT_READY",
                    message: "Card view not initialized",
                    details: nil
                ))
                return
            }
            cardView.getSessionData(result: result)
            
        // ==================== APPLE PAY METHODS ====================
        case "initApplePay":
            // Apple Pay initialization happens in PlatformView
            result(true)
            
        case "checkApplePayAvailability":
            guard let applePayView = applePayPlatformView else {
                result(false)
                return
            }
            
            Task {
                let isAvailable = await applePayView.checkAvailability()
                await MainActor.run {
                    result(isAvailable)
                }
            }
            
        case "getApplePaySessionData":
            // Session data is automatically sent via handleSubmit callback
            result(true)
            
        case "tokenizeApplePay":
            guard let applePayView = applePayPlatformView else {
                result(FlutterError(
                    code: "APPLEPAY_NOT_READY",
                    message: "Apple Pay view not initialized",
                    details: nil
                ))
                return
            }
            applePayView.tokenizeApplePay(result: result)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // Method to set card platform view reference
    func setCardPlatformView(_ view: CardPlatformView) {
        cardPlatformView = view
    }
    
    // Method to set Apple Pay platform view reference
    func setApplePayPlatformView(_ view: ApplePayPlatformView) {
        applePayPlatformView = view
    }
}
